text/trace/0205-messaging-semantic-conventions-context-propagation.md

```diff
@@ -0,0 +1,175 @@
+  +----------+
+       |
+       | HTTP POST (publishing a message)
+       v
++--------------+
+| Intermediary |
++--------------+
+       ^
+       | HTTP GET (fetching a message)
+       |
+  +----------+
+  | Consumer |
+  +----------+
+```
+
+Existing semantic conventions suppose that the consumer can use context
+information from the producer trace to create links or parent/child
+relationships between consumer and producer traces. For this to be possible,
+context information from the producer needs to be propagated to the consumer.
+In the example outlined above, the consumers sends an HTTP GET request to the
+intermediary to fetch a message, the message is returned as part of the
+response. Via this HTTP request, context information can be propagated from the
+consumer to the intermediary, but not from the intermediary to the consumer.
+The consumer can obtain the necessary producer context information only if it
+is propagated as part of the message itself, independent of HTTP context
+propagation.
+
+For correlating producer and consumer traces without special intermediary
+instrumentation it is thus necessary to attach a producer context to the
+message so it can be extracted and used by the consumer, regardless of the
+contexts that are propagated on HTTP requests for publishing and fetching the
+message.
+
+Although OpenTelemetry semantic conventions cannot specify the exact mechanisms
+to achieve this for every intermediary and every protocol, this requirement
+must be clearly formulated, so that it can be implemented by protocols and
+instrumentations.
+
+## Proposed addition to the messaging semantic conventions
+
+### Context propagation
+
+A message may pass many different components and layers in one or more
+intermediaries when it is propagated from the producer to the consumer. It
+cannot be assumed, and in many cases, it is not even desired, that all those
+components and layers are instrumented and propagate context according to
+OpenTelemetry requirements.
+
+A _message creation context_ allows correlating the producer with the
+consumer(s) of a message, regardless of intermediary instrumentation. The
+message creation context is created by the producer and must be propagated to
+the consumer(s). It must not be altered by intermediaries.  This context helps
+to model dependencies between producers and consumers, regardless of the
+underlying messaging transport mechanism and its instrumentation.
+
+Instrumentors are required to instrument producer and consumer applications
+so that context is attached to messages and extracted from messages in a
+coordinated way. Future versions of these conventions might recommend [context propagation according to certain industry standards](#standards-for-context-propagation).
+
+### Requirements
+
+A producer SHOULD attach a message creation context to each message. The message creation context
+SHOULD be attached in a way so that it is not possible to be changed by intermediaries.
```

Comment:
I would also mention that attaching a context per message is not always possible technically (redis pub sub, socket.io etc), and in those cases, there will be no correlation between producer and consumer. Although the SHOULD  word covers this case, it is so common that I think it deserves to be explicitly mentioned, and also communicate to readers that it is expected to happen and the implications.

Reply:
I added a remark about this in the section above.